### ISSUE TITLE
defense-finder won't find the all_best_solutions.tsv file

### ISSUE THREAD

**Issue by apcamargo** [mdmparis/defense-finder#5]
I'm trying to execute `defense-finder` with the following command:

```
defense-finder run --db-type gembase -o DEFENSE_FINDER_OUT -w 64 selected_scaffolds.faa
```

However, it seems that it can't find the `all_best_solutions.tsv` file:

```
cannot unpack non-iterable NoneType object
Traceback (most recent call last):
  File "/global/homes/a/acamargo/SCRATCH/.environments/defense/bin/defense-finder", line 8, in <module>
    sys.exit(cli())
  File "/global/homes/a/acamargo/SCRATCH/.environments/defense/lib/python3.7/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/global/homes/a/acamargo/SCRATCH/.environments/defense/lib/python3.7/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/global/homes/a/acamargo/SCRATCH/.environments/defense/lib/python3.7/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/global/homes/a/acamargo/SCRATCH/.environments/defense/lib/python3.7/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/global/homes/a/acamargo/SCRATCH/.environments/defense/lib/python3.7/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/global/homes/a/acamargo/SCRATCH/.environments/defense/lib/python3.7/site-packages/defense_finder_cli/main.py", line 46, in run
    defense_finder_posttreat.run(outdir)
  File "/global/homes/a/acamargo/SCRATCH/.environments/defense/lib/python3.7/site-packages/defense_finder_posttreat/__init__.py", line 7, in run
    bs = best_solutions.get()
  File "/global/homes/a/acamargo/SCRATCH/.environments/defense/lib/python3.7/site-packages/defense_finder_posttreat/best_solutions.py", line 4, in get
    parsed = parse_all()
  File "/global/homes/a/acamargo/SCRATCH/.environments/defense/lib/python3.7/site-packages/defense_finder_posttreat/best_solutions.py", line 8, in parse_all
    tsv_file = open('/tmp/defense-finder/all_best_solutions.tsv')
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/defense-finder/all_best_solutions.tsv'
```

The `/tmp/defense-finder` directory contains the following:
```
hmmer_results/
macsyfinder.conf
macsyfinder.log
```

**Comment by apcamargo** [comment]
This is most likely related to https://github.com/mdmparis/defense-finder/issues/3

**Comment by apcamargo** [comment]
After some investigation I found that the problem is that the MacSyFinder index was outdated after I changed my input file. Maybe `defense-finder` could call MacSyFinder using the `--force` parameter to avoid this kind of issue?

**Comment by pur80a** [comment]
I have the same issue: installed with pip 21.3.1 (python3.10.0) in an vitrualenv; to test I run: 
<pre>$ defense-finder run covid.fasta
...
FileNotFoundError: [Errno 2] No such file or directory: `'/tmp/defense-finder/all_best_solutions.tsv'</pre>
In `/tmp/defense-finder/macsyfinder.log` it says 
<pre>INFO     : command used: /apps/defensefinder/0.0.11/bin/macsyfinder --db-type ordered_replicon --sequence-db fastq_files/covid.fasta --models defense-finder-models all --out-dir /tmp/defense-finder --coverage-profile 0.1 --w 0</pre>
so I played around with `macsyfinder` and found out that I don't have no macsyfinder models in `<virtualenv>/share/macsyfinder/data/`, so I tried to see which models are available with `macsydata available` but got a SLL error. I then looked at [their website](https://github.com/gem-pasteur/macsyfinder) and saw that only python 3.7 | 3.8 | 3.9 are supported. So, I reinstalled defense-finder for python3.9.0. 
I could then check available models and download  `TXSS` and `TFF-SF`. 
I then rerun `defense-finder run covid.fasta` and still got `[Errno 2]`... 
when I run the `macsyfinde`r command mentioned above and delete `/tmp/defense-finder/macsyfinder.log`, I get: 
<pre>macsyfinder: "No such model definition: 'defense-finder-models'"</pre>

NOTE: I switched from legacy installation with setup.py to wheel without any success

**Comment by pur80a** [comment]
SOLUTION: 
<pre>pip install mdmparis-defense-finder==${VERSION}
defense-finder update
mv -v ~/.macsyfinder/data/defense-finder-models ${TARGETDIR}/share/macsyfinder/data/models
rm -rf ~/.macsyfinder</pre>
the solution was so obvious but rather hidden in the README.md file: use the update command after installation before using 
this could also be a solution for singularity #3 

I hope this helps people who encounter the same problem

**Comment by Alexandre-Herve** [comment]
Hi, sorry my notifications were incorrectly setup and I was unaware of your issues.

@apcamargo do you still have an issue?

**Comment by apcamargo** [comment]
I had two issues that were already solved:
- I forgot to delete the DefenseFinder index between two executions with different inputs (same file name). I solved this just by deleting the index and letting DefenseFinder create a new one. It would be nice to have a `--force` parameter to force DefenseFinder to create a new index.
- The other problem was actually [related to macsyfinder](https://github.com/gem-pasteur/macsyfinder/issues/42). I solved it by removing the lines of code that created a file that was not needed by DefenseFinder. Apparently this was fixed upstream, but I didn't test it yet.
